### PR TITLE
Add read/write support for time / datetime types

### LINKED ISSUE #98: Cannot write datetime column
```
In [7]: import pandas as pd

In [8]: import pyarrow as pa

In [9]: import lance

In [10]: df = pd.DataFrame({'date': pd.date_range('2022-01-01', periods=100, freq='D')})

In [11]: df.dtypes
Out[11]: 
date    datetime64[ns]
dtype: object

In [12]: table = pa.Table.from_pandas(df)

In [13]: table.schema
Out[13]: 
date: timestamp[ns]
-- schema metadata --
pandas: '{"index_columns": [{"kind": "range", "name": null, "start": 0, "' + 377

In [14]: lance.write_table(table, '/tmp/test.lance')
---------------------------------------------------------------------------
ArrowInvalid                              Traceback (most recent call last)
Input In [14], in <cell line: 1>()
----> 1 lance.write_table(table, '/tmp/test.lance')

File ~/code/eto/lance/python/lance/__init__.py:63, in write_table(table, destination)
     53 def write_table(table: pa.Table, destination: Union[str, Path]):
     54     """Write an Arrow Table into the destination.
     55 
     56     Parameters
   (...)
     61         The destination to write dataset to.
     62     """
---> 63     WriteTable(table, destination)

File ~/code/eto/lance/python/lance/_lib.pyx:117, in lance.lib.WriteTable()

File ~/code/eto/lance/python/thirdparty/arrow/python/pyarrow/error.pxi:100, in pyarrow.lib.check_status()

ArrowInvalid: PlainEncoder:: does not support data type timestamp[ns]
> /home/chang/code/eto/lance/python/scripts/pyarrow/error.pxi(100)pyarrow.lib.check_status()

ipdb> 
```

### PR DESCRIPTION
**Pull request by eddyxu**
Add read / write support for

- Timestamp(unit)
- Time32 / Time64
- Date32 / Date64

Closes #98

### CODE REVIEW

File: cpp/src/lance/io/reader.cc [hunk 1]
```diff
@@ -297,6 +297,10 @@ ::arrow::Result<std::shared_ptr<::arrow::Array>> FileReader::GetArray(
     storage_arr = GetDictionaryArray(field, batch_id, params);
   } else {
     storage_arr = GetPrimitiveArray(field, batch_id, params);
+    if (!storage_arr.ok()) {
+      return storage_arr.status();
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.cc:301]: how come in #104 we return `storage_arr` and here we return `storage_arr.status()` ?

File: cpp/src/lance/io/writer.cc [hunk 2]
```diff
@@ -108,7 +108,28 @@ ::arrow::Status FileWriter::WritePrimitiveArray(const std::shared_ptr<format::Fi
...
   auto encoder = field->GetEncoder(destination_);
-  ARROW_ASSIGN_OR_RAISE(auto pos, encoder->Write(arr));
+  auto type = field->type();
+
+  ::arrow::Result<std::shared_ptr<::arrow::Array>> physical_arr;
+  switch (type->id()) {
+    case ::arrow::TimestampType::type_id:
+    case ::arrow::Date64Type::type_id:
+    case ::arrow::Time64Type::type_id:
+      physical_arr = arr->View(::arrow::int64());
```
Comment (changhiskhan) [inline cpp/src/lance/io/writer.cc:118]: nit: is physical_arr just storage_arr? or is the concept here different?

File: cpp/src/lance/io/writer.cc [hunk 3]
```diff
@@ -108,7 +108,28 @@ ::arrow::Status FileWriter::WritePrimitiveArray(const std::shared_ptr<format::Fi
                                                 const std::shared_ptr<::arrow::Array>& arr) {
   auto field_id = field->id();
   auto encoder = field->GetEncoder(destination_);
-  ARROW_ASSIGN_OR_RAISE(auto pos, encoder->Write(arr));
+  auto type = field->type();
+
+  ::arrow::Result<std::shared_ptr<::arrow::Array>> physical_arr;
+  switch (type->id()) {
+    case ::arrow::TimestampType::type_id:
```
Comment (changhiskhan) [inline cpp/src/lance/io/writer.cc:115]: would it make sense to create methods like "lance::arrow::is_datetime64_type" and "lance::arrow::is_datetime32_type" or something like that?

**Review by changhiskhan — APPROVED** [review]
a few nits

File: cpp/src/lance/io/reader.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.cc:301]: `storage_arr` is a `Result<Array>`. While return `storage_arr` has the same effect as returning `storage_arr.status()`. Here is for more clarity w.r.t. code reading. 


Both ways are fine to me tho.

File: cpp/src/lance/io/writer.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/io/writer.cc:118]: Yes, they are the same. I can change the name to `storage_arr` .

**Pull request merged**